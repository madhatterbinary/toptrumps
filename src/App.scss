body {
  color: #000000;
  margin: 0 auto;
  padding: 1em;
  width: 95vw;
  max-width: 950px;
  font-weight: normal;
  font-size: 18px;
  background: #000000 url('images/wallpaper-starwars.png') no-repeat 1px 0;
    
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  height: 100vh;
}
.star-wars-top-trumps {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  text-align: center;
}

.control-nav,
.score-nav {
  flex: auto;
}

.score-nav {
  flex-basis: 8%;
  margin-left: auto;
  margin-right: auto;
  width: 440px;
  position: relative;
  top: 65px;
  z-index: 4;
}

.player-scores {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 0 10px;
}

.player-scores div {
  width: 100px;
  height: 100px;
  background: rgba(0, 0, 0, 0) url('images/deck1.png') no-repeat 0 0;
  background-size: contain;
  position: relative;
}

.player-score {
  color: black;
  font-weight: bold;
  padding: 10px;
  position: absolute;
  top: 30px;
  left: 45px;
}

.player-scores div:nth-child(even) {
  -moz-transform: scaleX(-1);
  -o-transform: scaleX(-1);
  -webkit-transform: scaleX(-1);
  transform: scaleX(-1);
  filter: FlipH;
  -ms-filter: "FlipH";
}

.player-scores div:nth-child(even) .player-score {
  -moz-transform: scaleX(-1);
  -o-transform: scaleX(-1);
  -webkit-transform: scaleX(-1);
  transform: scaleX(-1);
  filter: FlipH;
  -ms-filter: "FlipH";
}

.game-panel {
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex-grow: 1;
  padding: 0 50px;
}

.game-nav-result {
  color: white;
  background-color: #65ee0a;
  color: #232323;;
  border-radius: 5px;
  font-size: 2em;
  margin-top:210px;
  margin-bottom: 10px;
  visibility: visible;
  position: absolute;
  padding: 0 80px;
  z-index: 2;
}

.card-container {
  padding: 10px;
}

.card-container.winner .card-result {
  visibility: visible;
}

.card-result {
  color: #28a745;
  border-radius: 5px;
  font-size: 2em;
  margin-top:0px;
  margin-bottom: 10px;
  visibility: hidden;
  margin-left: 20px;
  margin-right: 20px;
}

.card {
  background: url('images/cardfront.png') no-repeat 0 -5px;
  width: 300px;
  height: 456px;
  padding: 10px;
  background-size: cover;
}

.card.closed {
  background: url('images/cardback.png') no-repeat 0 -5px;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  height: 460px;
  background-size: cover;
}

.card-header {
  background-color: #343444;
  color: white;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding: 5px;
}

.card-header h3 {
  text-align: left;
  padding: 0;
  margin: 0 0 7px 0;
  display: block;
  width: 250px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-left: 10px;
  padding-right: 10px;
}

.card-additional-info {
  overflow: hidden;
  max-width: 100%;
  padding-left: 10px;
  padding-right: 10px;
}

.card-additional-info div{
  text-align: left;
  font-size: .7em;
  font-style: italic;
  margin: 0 0 5px 0;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.card-image-container {
  overflow: hidden;
  border: 1px solid #232323;
  background-color: #000;
}

.card-image {
  height: auto;
  display: block;
  margin: 0 auto;
  width: 260px;
  max-height: 134px;
}
.card-body {
  padding: 0.5rem;
  padding-top: 45px;
  border:  #333333 solid 1px;
    border-radius: 12px;
}

.card-attribute-list {
  border: 1px solid #232323;
  border-top: 0;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  background-color: #232323;
  padding: 0 0 5px 0;
  font-size: .8em;
  color: white;
}

.card-attribute-row:nth-child(odd) {
  background-color: #343444;
}

.card-attribute-row {
  padding: 2px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding-left: 10px;
  padding-right: 10px;
  cursor: pointer;
  opacity: 1;
  pointer-events:all;
}

.card-attribute-row.disabled {
  pointer-events: none;
  opacity: 0.3;
}
.card-attribute-row:hover {
  background-color: #16161b;
}

.card-attribute-value {
  cursor: pointer;
}

.card-attribute-value:hover,
.card-attribute-value.selected,
.card-attribute-label.selected {
  color: #65ee0a;
}

.selected-attribute {
  margin-bottom: 10px;
  color: #65ee0a;
}

.control-nav {
  padding-top: 20px;
  display: block;
  bottom: 50px;
  position: relative;
}

.button {
  width: 200px;
}

/* start da css for da buttons */
.btn {
  border-radius: 5px;
  padding-bottom: 40px;
  padding-top: 10px;
  width: 150px;
  font-size: 22px;
  text-decoration: none;
  margin: 20px;
  color: #fff;
  position: relative;
  display: inline-block;
  cursor: pointer;
  background-color:  #1d1d27;
  height: 50px;
}

.btn.dark.disabled {
  background-color: #333;
  color: #666;
  cursor: not-allowed;
}

.btn:active {
  transform: translate(0px, 5px);
  -webkit-transform: translate(0px, 5px);
  box-shadow: 0px 1px 0px 0px;
}
.dark {
  background-color: #green;
}

.dark:hover {
  background-color: #blue;
}

.top-trump {
  background-color:  #1d1d27;
  color: white;
  cursor: pointer;
}

.top-trump:hover, .btn:hover {
  background-color: #343444;
  color: white;
}
.winner-box {
  color: white;
  font-size:12px;
  span {
    font-size:24px;
    color: #28a745;
  }
}
.badge {
  font-size: 100%;
}
