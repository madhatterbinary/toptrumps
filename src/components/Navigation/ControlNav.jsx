import React from 'react';
import PropTypes from 'prop-types';
import Button from '../common/Button';
import { STATUS_READY, STATUS_OVER } from '../../constants/constants';

const ControlNav = (props) => { 
  const {
    status,
    isPlayEnabled,
    handleButtonClick,
  } = props;

  let buttonLabel;
  switch(status) {
    case STATUS_READY:
      buttonLabel = 'Play';
      break;
    case STATUS_OVER:
      buttonLabel = 'Play Again';
      break;
    default:
      buttonLabel = 'Shuffle';
  }

  return (
    <div className="control-nav">
      <Button
        buttonClassName={status === STATUS_READY ? 'dark' : 'top-trump'}
        buttonOnClick={handleButtonClick}
        buttonLabel={buttonLabel}
        buttonIsDisabled={!isPlayEnabled}
      />
    </div>
  );
}

ControlNav.propTypes = {
  status: PropTypes.string.isRequired,
  isPlayEnabled: PropTypes.bool.isRequired,
  handleButtonClick: PropTypes.func.isRequired,
};

export default ControlNav;
