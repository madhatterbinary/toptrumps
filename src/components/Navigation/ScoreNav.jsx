import React from 'react';
import PropTypes from 'prop-types';

const ScoreNav = ({ player1Score, player2Score }) => (
  <div className="score-nav">
    <div className="player-scores">
      <div><span className="player-score">{player1Score}</span></div>
      <div><span className="player-score">{player2Score}</span></div>
    </div>
  </div>
);

ScoreNav.propTypes = {
  player1Score: PropTypes.number,
  player2Score: PropTypes.number
};

export default ScoreNav;
