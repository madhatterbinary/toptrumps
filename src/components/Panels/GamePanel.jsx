import React from 'react';
import PropTypes from 'prop-types';
import CardContainer from '../card/CardContainer';
import {
  STATUS_READY,
  STATUS_OVER,
  PLAYER_1,
  PLAYER_2,
} from '../../constants/constants';

const GamePanel = (props) => {
  const {
    player1Card,
    player2Card,
    status,
    handleAttributeSelection,
    selectedAttribute,
    lastWinner,
    isDraw,
  } = props;

  return (
    <div className="game-panel">
      { status === STATUS_OVER && (
        <div className="game-nav-result">Game Over</div>
      )}
      { isDraw && (
        <div className="game-nav-result">Draw</div>
      )}
      <CardContainer
        card={player1Card}
        isClosed={status === STATUS_READY && lastWinner === PLAYER_2}
        handleAttributeSelection={handleAttributeSelection}
        selectedAttribute={selectedAttribute}
        status={status}
        isWinner={status !== STATUS_READY && lastWinner === PLAYER_1 && !isDraw}
      />
      <CardContainer 
        card={player2Card}
        isClosed={status === STATUS_READY && lastWinner !== PLAYER_2}
        handleAttributeSelection={handleAttributeSelection}
        selectedAttribute={selectedAttribute}
        status={status}
        isWinner={status !== STATUS_READY && lastWinner === PLAYER_2 && !isDraw}
      />
    </div>
  );
}

GamePanel.propTypes = {
  status: PropTypes.string.isRequired,
  player1Card: PropTypes.object,
  player2Card: PropTypes.object,
  lastWinner: PropTypes.string,
  selectedAttribute: PropTypes.string,
  handleAttributeSelection: PropTypes.func.isRequired,
}

export default GamePanel;
