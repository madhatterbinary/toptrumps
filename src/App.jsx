import React, { useEffect, useState, useCallback, Fragment } from 'react';
import ScoreNav from './components/Navigation/ScoreNav';
import GamePanel from './components/Panels/GamePanel';
import ControlNav from './components/Navigation/ControlNav';
import shuffleArray from './utils/shuffleArray';
import ImagesData from './data/images.json';
import useSWR from "swr";
import {
  STATUS_READY,
  STATUS_DONE,
  STATUS_OVER,
  STATUS_REPLAY,
  PLAYER_1,
  PLAYER_2,
} from './constants/constants';
import './App.scss';

const url = "https://swapi.dev/api/vehicles/";

function App () {
  const [game, setGame] = useState({firstRender: true});
  const [errorMsg, setErrorMsg] = useState("");
  const [loadingMsg, setLoadingMsg] = useState("");
  const fetcher = (...args) => fetch(...args).then((res) => res.json());
  const { data: result, error } = useSWR(url, fetcher);

  const handleGameOver = useCallback(() => {
    if(game.status === "over" && !game.firstRender) {
      setGame({...game, status: STATUS_REPLAY});
    }
  },[game]);

  const handleResetGame = useCallback(() => {
    if (game.firstRender || game.status === STATUS_REPLAY) {
      if (error) {
        setErrorMsg("Something went wrong");
      } else if (!result) {
        setLoadingMsg("Loading...");
      } else {
        const images = Object.values(ImagesData.images);
        const resultAddedImgs = result.results.map((item,index) => {
          return {...item, image:Object.values(images[index])[0]};
        });
        const results = shuffleArray(resultAddedImgs);
        const halfLength = Math.ceil(results.length / 2);
        setGame({
          status: STATUS_READY,
          lastWinner: undefined,
          deck1: results.slice(0, halfLength),
          deck2: results.slice(halfLength),
          selectedAttribute: undefined,
          isDraw: false,
          firstRender: false,
          winner: undefined,
        });
      }
    }
  }, [game, result, error]);

  useEffect(() =>{
     handleResetGame();
  },[handleResetGame, game, game.winner]);

  const handleAttributeSelection = (attributeKey) => {
    setGame({
      ...game,
      selectedAttribute: attributeKey
    })
  }

  const handleButtonClick= () => {
    const card1 = game.deck1[0];
    const card2 = game.deck2[0];
    switch (game.status) {
      case STATUS_OVER:
        handleGameOver();
        break;
      case STATUS_REPLAY:
        handleResetGame();
        break;
      case STATUS_READY:
        handlePlay(card1, card2);
        break;
      default:
        handleShuffle(card1, card2);
    }
  }

  const willGameEnd = (game) => {
    return !game.isDraw && 
      ((game.lastWinner === PLAYER_1 && game.deck2.length === 1) ||
      (game.lastWinner === PLAYER_2 && game.deck1.length === 1));
  }

  const handlePlay = (card1, card2) => {
    const attribute1 = parseInt(card1[game.selectedAttribute], 10);
    const attribute2 = parseInt(card2[game.selectedAttribute], 10);
    let newGame = {};

    if (card1.topTrump || attribute1 > attribute2) {
      newGame = {...game, lastWinner: PLAYER_1 };
    } else if (card2.topTrump || attribute1 < attribute2) {
      newGame = {...game, lastWinner: PLAYER_2 };
    } else {
      newGame = {...game, isDraw: true };
    }
    newGame = {...newGame,
     status: willGameEnd(newGame) ? STATUS_OVER : STATUS_DONE,
     winner: willGameEnd(newGame) ? newGame.lastWinner : undefined
    };
    setGame({
      ...newGame
      });
  }

  const handleShuffle= (card1, card2) => {
    const newGame = {
      ...game,
      status: STATUS_READY,
      selectedAttribute: undefined,
      isDraw: false,
    };
    if (game.isDraw) {
      newGame.deck1 = [...game.deck1.slice(1), card1];
      newGame.deck2 = [...game.deck2.slice(1), card2];
    } else {
      newGame.deck1 = [...game.deck1.slice(1), ...(game.lastWinner === PLAYER_1 ? [card1, card2] : [])];
      newGame.deck2 = [...game.deck2.slice(1), ...(game.lastWinner === PLAYER_2 ? [card1, card2] : [])];
    }
    setGame({
      ...game,
      ...newGame
    });
  }
    if (error) {
      return <div className="star-wars-top-trumps">{errorMsg}</div>;
    } else if (!result) {
      return <div className="star-wars-top-trumps">{loadingMsg}</div>
    }

    return (
      <div className="star-wars-top-trumps">
      {
        game && game.deck1 && (
        <Fragment>
          <ScoreNav
            player1Score={
              game.status === "over" && game.deck1.length === 1 ?
              0 :
              game.status === "over" && game.deck1.length === 9 ?
              10 :
              game.deck1.length
            }
            player2Score={
              game.status === "over" && game.deck2.length === 1 ?
              0 :
              game.status === "over" && game.deck2.length === 9 ?
              10 :
              game.deck2.length
            }
          />
          {game.status === STATUS_OVER && (
            <div className={`winner-box ${game.winner}`}>
            {game.winner ==="player1" ? 
            <span>Player 1 won</span> :
            <span>Player 2 won</span>
            }
            </div>
          )}
          <GamePanel
            player1Card={game.deck1[0]}
            player2Card={game.deck2[0]}
            selectedAttribute={game.selectedAttribute}
            status={game.status}
            handleAttributeSelection={handleAttributeSelection}
            lastWinner={game.lastWinner}
            isDraw={game.isDraw}
          />
          <ControlNav
            status={game.status}
            isPlayEnabled={!!game.selectedAttribute}
            handleButtonClick={handleButtonClick}
          />
        </Fragment>
        )
      }
      </div>
    );
}

export default App;
