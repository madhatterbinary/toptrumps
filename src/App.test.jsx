import React from 'react';
import App from './App';
import ScoreNav from './components/Navigation/ScoreNav';
import GamePanel from './components/Panels/GamePanel';
import ControlNav from './components/Navigation/ControlNav';

import TestRenderer from 'react-test-renderer';

describe('App', () => {
  it('renders', () => {
    const component = TestRenderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(component.root.findAllByType(ScoreNav)).not.toBeNull();
    expect(component.root.findAllByType(GamePanel)).not.toBeNull();
    expect(component.root.findAllByType(ControlNav)).not.toBeNull();
  });
});
