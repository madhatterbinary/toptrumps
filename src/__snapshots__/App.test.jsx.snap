// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div
  className="star-wars-top-trumps"
>
  
</div>
`;
